libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -134,8 +134,46 @@ string IRGenerationContext::trySuccessConditionVariable(Expression const& _expre
 
 string IRGenerationContext::internalDispatch(Arity const& _arity)
 {
-	string funName = "dispatch_internal_in_" + to_string(_arity.in) + "_out_" + to_string(_arity.out);
-	return m_functions.createFunction(funName, [&]() {
+	// UNIMPLEMENTED: Internal library calls via pointers are not implemented yet.
+	// We're not generating code for internal library functions here even though it's possible
+	// to call them via pointers. Right now such calls end up triggering the `default` case in
+	// the switch above.
+	set<FunctionDefinition const*> functions;
+	for (auto const& contract: mostDerivedContract().annotation().linearizedBaseContracts)
+		for (FunctionDefinition const* function: contract->definedFunctions())
+			if (
+				FunctionType const* functionType = TypeProvider::function(*function)->asCallableFunction(false);
+				!function->isConstructor() &&
+				TupleType(functionType->parameterTypes()).sizeOnStack() == _arity.in &&
+				TupleType(functionType->returnParameterTypes()).sizeOnStack() == _arity.out
+			)
+			{
+				functions.insert(function);
+				enqueueFunctionForCodeGeneration(*function);
+			}
+
+	return internalDispatch(_arity, functions);
+}
+
+string IRGenerationContext::internalDispatch(Arity const& _arity, set<FunctionDefinition const*> const& _functions)
+{
+	vector<map<string, string>> cases;
+	for (auto const* function: _functions)
+	{
+		solAssert(function, "");
+		solAssert(getFunctionArity(*function) == _arity, "A single dispatch function can only handle functions of one arity");
+		solAssert(!function->isConstructor(), "");
+		// 0 is reserved for uninitialized function pointers
+		solAssert(function->id() != 0, "Unexpected function ID: 0");
+
+		cases.emplace_back(map<string, string>{
+			{"funID", to_string(function->id())},
+			{"name", buildFunctionName(*function)}
+		});
+	}
+
+	string funName = buildInternalDispatchFunctionName(_arity);
```

Comment:
I would be happier if somehow the IDs of the functions could be included in the function name.

Reply:
There could be a lot of them. Surely Yul has a limit on function name length?

Anyway, if it's doable I can do it but I think this would be a separate PR. Doesn't seem hard.

Reply:
We could hash them? I'm concerned about this function being called twice with different function definitions and then not doing anything on the second call.

Reply:
That's how the function was designed and this PR is not changing that. It's meant to be called for every call via pointer and to generate the dispatch only the first time (because it includes all the functions from the contract in it anyway).

#8797 changes that and I added an assertion there according to your suggestion. There `generateInternalDispatchFunction()` can no longer be called multiple times with the same arities.

Reply:
So adding a hash might still be useful just to be extra safe but the problem should no longer exist in #8797.

Reply:
ok!

Reply:
So should I create a PR that adds the hash or is the assertion enough?

Reply:
Another difficulty here is that you need to know the name before you find all the dispatchable functions. So including a hash of the IDs would require a second pass through the generated code to update the name.